clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -637,6 +638,8 @@ private void handleResponse(RequestHeader header, Struct body, long now) {
                 log.trace("Ignoring empty metadata response with correlation id {}.", header.correlationId());
                 this.metadata.failedUpdate(now);
             }
+            for (String topic : response.errorTopics(Errors.UNKNOWN_TOPIC_OR_PARTITION))
+                this.metadata.handleUnknownTopic(topic);
```

Comment:
@ewencp I'm not sure I understand. The point of the cache is to avoid unnecessary metadata fetches. If we remove the entry and the user continues trying to access it, then we'll just have to fetch it again. Wouldn't it be better to keep the cached error as long as the user is still trying to write to the topic?